Removed TempFile and TempDirectory from FileSystemWatcher Tests

Danyy427 opened the pull request:
relates to #68430

Commit:
Removed TempFile and TempDirectory from FileSystemWatcher Tests

Commit:
Added missing CreateDirectory

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Changed.cs:
```diff
@@ -14,28 +14,30 @@ public void FileSystemWatcher_Directory_Changed_LastWrite()
-                using (var testDirectory = new TempDirectory(GetTestFilePath()))
-                using (var dir = new TempDirectory(Path.Combine(testDirectory.Path, "dir")))
-                using (var watcher = new FileSystemWatcher(testDirectory.Path, Path.GetFileName(dir.Path)))
+                string testDirectory = TestDirectory;
```
you don't need a variable here, just use TestDirectory.

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Changed.cs:
```diff
@@ -44,57 +46,64 @@ public void FileSystemWatcher_Directory_Changed_WatchedFolder()
+            string dir = TestDirectory;
+            string firstDir = Path.Combine(dir, "dir1");
+            string nestedDir = Path.Combine(firstDir, "nested");
+            Directory.CreateDirectory(firstDir);
```
creating a nested dir will automatically create any necessary parent dirs, so you do'nt need to create them first.

danmoseley commented:
Thanks. couple things I noticed -- if you can fix those, and verify tests pass locally for you, I can have another look.

Danyy427 commented:
@danmoseley I did that variable thing everywhere so I'll have to replace them all but that won't take much time probably.

Commit:
Removed redundant variables

Danyy427 commented:
@danmoseley that should be all

danmoseley commented:
do these tests pass locally?

Danyy427 commented:
@danmoseley yes they do

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Changed.cs:
```diff
@@ -44,57 +44,59 @@ public void FileSystemWatcher_Directory_Changed_WatchedFolder()
-            using (var firstDir = new TempDirectory(Path.Combine(dir.Path, "dir1")))
-            using (var nestedDir = new TempDirectory(Path.Combine(firstDir.Path, "nested")))
-            using (var watcher = new FileSystemWatcher(dir.Path, "*"))
+            string firstDir = Path.Combine(TestDirectory, "dir1");
+            string nestedDir = Path.Combine(firstDir, "nested");
```
```suggestion
            string nestedDir = Path.Combine(firstDir, "dir1", "nested");
```

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Changed.cs:
```diff
@@ -44,57 +44,59 @@ public void FileSystemWatcher_Directory_Changed_WatchedFolder()
-            using (var file = new TempFile(Path.Combine(tempDir.Path, "test")))
-            using (var watcher = new FileSystemWatcher(dir.Path, "*"))
+            string dir = Path.Combine(TestDirectory, "dir");
+            string tempDir = Path.Combine(dir, "tempDir");
+            string file = Path.Combine(tempDir, "test");
+            Directory.CreateDirectory(tempDir);
```
```suggestion
            Directory.CreateDirectory(Path.Combine(dir, "tempDir"));
            string file = Path.Combine(dir, "tempDir", "test");
            
```
is that cleaner? up to you

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Create.cs:
```diff
@@ -49,15 +48,15 @@ public void FileSystemWatcher_Directory_Create()
-            using (var firstDir = new TempDirectory(Path.Combine(dir.Path, "dir1")))
-            using (var nestedDir = new TempDirectory(Path.Combine(firstDir.Path, "nested")))
-            using (var watcher = new FileSystemWatcher(dir.Path, "*"))
+            string firstDir = Path.Combine(TestDirectory, "dir1");
+            string nestedDir = Path.Combine(firstDir, "nested");
```
```suggestion
            string nestedDir = Path.Combine(firstDir, "dir1", "nested");
```
and others below similarly

Commit:
removed redundant variable

Co-authored-by: Dan Moseley <danmose@microsoft.com>

Commit:
removed redundant variable

Co-authored-by: Dan Moseley <danmose@microsoft.com>

Commit:
removed redundant variable

Co-authored-by: Dan Moseley <danmose@microsoft.com>

danmoseley commented:
I could imagine adding something like these to FileCleanupTestBase
```c#
protected string CreateDirectory(params string[] paths)
{
        string dir = Path.Combine(paths);
        Directory.CreateDirectory(dir);
        return dir;
}

protected string GetTestFile()
{
       string file = GetTestFilePath();
       File.Create(file).Dispose();
       return file;
}
```
then code like
```c#
            string dir = Path.Combine(TestDirectory, "dir");
            string temp = GetTestFilePath();
            Directory.CreateDirectory(dir);
            File.Create(temp).Dispose();
```
could become just
```c#
         string dir = CreateDirectory(TestDirectory, "dir");
         string temp = GetTestFile();
```
would that be worthwhile?

Danyy427 commented:
@danmoseley Yes, I thought about the same thing.

danmoseley commented:
edited above. Also CreateDirectory and CreateTestFile would probably implicitly start in TestDirectory, so you wouln't need to pass that in.

Danyy427 commented:
@danmoseley Should I add those methods?

danmoseley commented:
@Danyy427 if you think something like that would be cleaner, sure!

Danyy427 commented:
@danmoseley we cannot do 

```
string dir = Path.Combine(TestDirectory, paths);
```
there is no overload of `Combine` that takes a `string` and a `string[]`. Should I do it with two `Combine`'s or do you know of a better way?

Danyy427 commented:
@danmoseley also the build failed from FleSystemWatcher Tests, I think we messed something with those last commits

Commit:
Fixed directory name

Commit:
Reverted changes

Danyy427 commented:
@danmoseley on one of them `firstDir` had to be replaced with `TestDirectory`, I reverted the other one because the deleted variable was being used inside the method

Danyy427 commented:
@danmoseley adding `TestDirectory` to `CreateTestDirectory` makes it harder to use. I made it so that it only combines the parameters and `TestDirectory` has to be added manually

Commit:
Simplified

Danyy427 commented:
@danmoseley added the methods, the local tests pass

Commit:
made local tests

Danyy427 commented:
@danmoseley  It seems like local tests didn't actually passed :/

Commit:
substituted methods

danmoseley commented on src/libraries/Common/tests/TestUtilities/System/IO/FileCleanupTestBase.cs:
```diff
@@ -214,5 +214,25 @@ private unsafe uint GetFinalPathNameByHandle(SafeFileHandle handle, char[] buffe
+        protected string CreateTestFile(params string[] paths)
+        {
+            string file = Path.Combine(paths);
+            File.Create(file).Dispose();
```
```suggestion
            Assert.True(Path.IsPathRooted(file));
            Directory.CreateDirectory(Path.GetDirectoryName(file));
            File.Create(file).Dispose();
```
that way you don't have to create any directory first.

Danyy427 replied:
I think we will have to create the directories anyways since their paths are used inside most of the methods anyways.

danmoseley replied:
Not all of them, eg I pointed out this one
```
            string firstDir = Path.Combine(TestDirectory, "dir1");
            string nestedDir = CreateTestDirectory(firstDir, "nested");
            string nestedFile = CreateTestFile(nestedDir, "nestedFile");
```

danmoseley commented on src/libraries/Common/tests/TestUtilities/System/IO/FileCleanupTestBase.cs:
```diff
@@ -214,5 +214,25 @@ private unsafe uint GetFinalPathNameByHandle(SafeFileHandle handle, char[] buffe
+        protected string CreateTestDirectory(params string[] paths)
+        {
+            string dir = Path.Combine(paths);
+            Directory.CreateDirectory(dir);
```
```suggestion
            Assert.True(Path.IsPathRooted(file));
            Directory.CreateDirectory(dir);
```
this ensures we're not depending on whatever the current directory is -- the path is rooted.

danmoseley commented on src/libraries/Common/tests/TestUtilities/System/IO/FileCleanupTestBase.cs:
```diff
@@ -214,5 +214,25 @@ private unsafe uint GetFinalPathNameByHandle(SafeFileHandle handle, char[] buffe
+            return dir;
+        }
+
+        protected string CreateTestDirectory()
+             => CreateTestDirectory(GetTestFilePath());
```
```suggestion
        protected string CreateTestDirectory() => CreateTestDirectory(GetTestFilePath());
```

danmoseley commented on src/libraries/Common/tests/TestUtilities/System/IO/FileCleanupTestBase.cs:
```diff
@@ -214,5 +214,25 @@ private unsafe uint GetFinalPathNameByHandle(SafeFileHandle handle, char[] buffe
+            File.Create(file).Dispose();
+            return file;
+        }
+        protected string CreateTestFile()
```
nit, missing blank line above this method.

danmoseley commented on src/libraries/Common/tests/TestUtilities/System/IO/FileCleanupTestBase.cs:
```diff
@@ -214,5 +214,25 @@ private unsafe uint GetFinalPathNameByHandle(SafeFileHandle handle, char[] buffe
+            File.Create(file).Dispose();
+            return file;
+        }
+        protected string CreateTestFile()
+            => CreateTestFile(GetTestFilePath());
```
```suggestion
        protected string CreateTestFile() => CreateTestFile(GetTestFilePath());
```

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Changed.cs:
```diff
@@ -27,32 +26,30 @@ public void FileSystemWatcher_File_Changed_LastWrite()
-            using (var nestedDir = new TempDirectory(Path.Combine(firstDir.Path, "nested")))
-            using (var nestedFile = new TempFile(Path.Combine(nestedDir.Path, "nestedFile")))
-            using (var watcher = new FileSystemWatcher(dir.Path, "*"))
+            string firstDir = Path.Combine(TestDirectory, "dir1");
+            string nestedDir = CreateTestDirectory(firstDir, "nested");
+            string nestedFile = CreateTestFile(nestedDir, "nestedFile");
```
```suggestion
            string nestedFile = CreateTestFile(TestDirectory, "dir1", "nested", "nestedFile");
```
firstDir, nestedDir are not used, and we made CreateTestFile create any directories it needs.

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Changed.cs:
```diff
@@ -67,34 +64,33 @@ public void FileSystemWatcher_File_Changed_DataModification()
-            using (var file = new TempFile(Path.Combine(testDirectory.Path, "file")))
-            using (var watcher = new FileSystemWatcher(testDirectory.Path, Path.GetFileName(file.Path)))
+            string file = CreateTestFile(TestDirectory, "file");
+            File.Create(file).Dispose();
```
```suggestion
```
you don't need this any more. I see this also in 3 other places.

Danyy427 replied:
I couldn't find any more of these.

Commit:
fix

Co-authored-by: Dan Moseley <danmose@microsoft.com>

Commit:
Update src/libraries/Common/tests/TestUtilities/System/IO/FileCleanupTestBase.cs

spacing

Co-authored-by: Dan Moseley <danmose@microsoft.com>

Commit:
spacing

Co-authored-by: Dan Moseley <danmose@microsoft.com>

Commit:
remove redundant file creation

Co-authored-by: Dan Moseley <danmose@microsoft.com>

danmoseley commented:
@Danyy427 what we normally do when someone's reviewing is not push more changes until we think they're done. All that does is restart all the validation machines, which is just burns electricity as they'll probably have more comments. 😄  Also, it's less confusing. 

You'll want to apply any comments (assuming their correct..) to similar issues elsewhere (people don't always point out every instance). Also, you'll want to re-run tests locally before pushing.

Danyy427 commented:
@danmoseley oops, sorry about that. I have applied the suggestions as necessary and commented on those I am not so sure about.

Commit:
fixes

Danyy427 commented:
@danmoseley I modified `CreateTestFile` and `CreateTestDirectory` as needed. Replaced the instances where directory names are not used within the method with `CreateTestFile`, and added the missing blank line. The local tests do pass.

Commit:
fix

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Delete.cs:
```diff
@@ -29,15 +28,14 @@ public void FileSystemWatcher_Directory_Delete()
-            using (var firstDir = new TempDirectory(Path.Combine(dir.Path, "dir1")))
-            using (var nestedDir = new TempDirectory(Path.Combine(firstDir.Path, "nested")))
-            using (var watcher = new FileSystemWatcher(dir.Path, "*"))
+            string firstDir = Path.Combine(TestDirectory, "dir1");
+            string nestedDir = CreateTestDirectory(firstDir, "nested");
```
```suggestion
            string nestedDir = CreateTestDirectory(TestDirectory, "dir1", "nested");
```

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Delete.cs:
```diff
@@ -70,15 +67,14 @@ public void FileSystemWatcher_Directory_Delete_DeepDirectoryStructure()
-            using (var tempDir = new TempDirectory(GetTestFilePath()))
-            using (var watcher = new FileSystemWatcher(Path.GetFullPath(dir.Path), "*"))
+            string dir = CreateTestDirectory(TestDirectory, "dir");
+            string tempDir = CreateTestFile();
```
```suggestion
            string tempDir = CreateTestDirectory(TestDirectory, "tempDir");
```

danmoseley replied:
have you tried out running tests in WSL2 -- so you can run tests on Linux locally? This will fail there (ought to fail on Windows too though)

Danyy427 replied:
I haven't tried running them in WSL. They did not fail on Windows though

Danyy427 replied:
@danmoseley instead of `CreateTestDirectory(TestDirectory, "tempDir")`, I just used `CreateTestDirectory()` since the original code suggests `TempDirectory(GetTestFilePath())`

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.Directory.Move.cs:
```diff
@@ -192,20 +189,19 @@ private void DirectoryMove_Multiple_FromUnwatchedToWatched(int filesCount)
-            using var watchedTestDirectory = new TempDirectory(GetTestFilePath());
-            using var unwatchedTestDirectory = new TempDirectory(GetTestFilePath());
-
+            string watchedTestDirectory = CreateTestDirectory(GetTestFilePath());
```
replace all `CreateTestDirectory(GetTestFilePath())`  with `CreateTestDirectory()`

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.WaitForChanged.cs:
```diff
@@ -80,9 +80,8 @@ public static void WaitForChangedResult_TimedOut_Roundtrip()
-            using var testDirectory = new TempDirectory(GetTestFilePath());
-            using var _ = new TempDirectory(Path.Combine(testDirectory.Path, GetTestFileName()));
-            using var fsw = new FileSystemWatcher(testDirectory.Path);
+            string _ = Path.Combine(TestDirectory, GetTestFileName());
```
this code doesn't do anything. it should be `var _ = CreateTestDirectory();`

danmoseley commented on src/libraries/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.cs:
```diff
@@ -1122,30 +1100,29 @@ public void FileSystemWatcher_CreateManyConcurrentWatches()
                     for (int i = 0; i <= maxUserWatches; i++)
                     {
-                        Directory.CreateDirectory(Path.Combine(dir.Path, i.ToString()));
+                        Directory.CreateDirectory(Path.Combine(TestDirectory, i.ToString()));
```
```suggestion
                        CreateDirectory(TestDirectory, i.ToString());
```

Danyy427 replied:
@danmoseley  Do you mean `CreateTestDirectory` here? Also, do we need to add it here? We are not using any return values and therefore I think it would be redundant to use them here.

danmoseley commented:
added some more comments

Commit:
applied suggestions, removed redundant variables

Commit:
spacing

Danyy427 commented:
@danmoseley are these failures related to this PR. One of them seems to be related with ```FileSystemWatcher_SymbolicLink_IncludeSubdirectories_DoNotDereferenceChildLink```. I checked the code and we haven't changed anything from the original per se.

Danyy427 commented:
@danmoseley could we retry this build again? I am not so sure that the fails are related to the code (especially system.codedom ones). I couldn't find the retry build or something like that anywhere.

Danyy427 commented:
@danmoseley I was just able to test these at WSL. The only test that doesn't pass is `'System.Formats.Tar.Tests'` which I doubt is related to our issue. Do you think there are any more changes to be made?

danmoseley commented:
LGTM, but there is a relevant failure in FileSystemWatcher_SymbolicLink_IncludeSubdirectories_DoNotDereferenceChildLink. And running locally I get a differently -- consistently. I'm not certain why yet, but I suspect we have introduced a problem by putting files directly into the tests' folder. Looking...

Commit:
Fix casing bug

danmoseley commented:
I figured out my failure -- it took two hours 😢 

See https://github.com/dotnet/runtime/pull/68444/commits/a0d598d7b588c6524491f3ac695e19f3c90e3a3c

This PR changed casing of the file to "otherFile" and watched for a change to "otherfile"

Note that this test was 'outerloop' so you did not see it locally. You need to pass an extra flag "/p:outerloop=true" to run those.

I pushed the fix and rebased.

Danyy427 commented:
@danmoseley Wow that's bad. I know the feeling.

danmoseley reviewed (approved):
thanks for the contribution @Danyy427

Danyy427 commented:
@danmoseley Happy if I could help.

danmoseley commented:
Verified outerloop passes on Windows and Linux.
